File: src/docforge/openapi/paths.py
import re
from collections import defaultdict


class SlugCollisionError(ValueError):
    """Two distinct (method, path) pairs map to the same output filename."""


_NON_SLUG_CHARS = re.compile(r"[/{}]+")
_MULTI_UNDERSCORE = re.compile(r"_+")


def slug_path(path: str) -> str:
    s = _NON_SLUG_CHARS.sub("_", path)
    s = _MULTI_UNDERSCORE.sub("_", s).strip("_")
    return s or "root"


def endpoint_filename(method: str, path: str) -> str:
    return f"{method.upper()}_{slug_path(path)}.md"


def schema_filename(name: str) -> str:
    return f"{name}.md"


def detect_endpoint_collisions(pairs: list[tuple[str, str]]) -> None:
    inverse: dict[str, list[tuple[str, str]]] = defaultdict(list)
    for method, path in pairs:
        inverse[endpoint_filename(method, path)].append((method, path))
    dupes = {k: v for k, v in inverse.items() if len(v) > 1}
    if dupes:
        lines = ["endpoint filename collisions:"]
        for fname, sources in dupes.items():
            lines.append(f"  -> {fname}")
            for m, p in sources:
                lines.append(f"      from {m.upper()} {p}")
        raise SlugCollisionError("\n".join(lines))

File: src/openapi/paths.ts
const NON_SLUG_CHARS = /[/{}]+/g;
const MULTI_UNDERSCORE = /_+/g;

export class SlugCollisionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SlugCollisionError";
  }
}

export function slugPath(path: string): string {
  let s = path.replace(NON_SLUG_CHARS, "_");
  s = s.replace(MULTI_UNDERSCORE, "_");
  s = s.replace(/^_+|_+$/g, "");
  return s || "root";
}

export function endpointFilename(method: string, path: string): string {
  return `${method.toUpperCase()}_${slugPath(path)}.md`;
}

export function schemaFilename(name: string): string {
  return `${name}.md`;
}

export function detectEndpointCollisions(
  pairs: Array<[string, string]>,
): void {
  const inverse = new Map<string, Array<[string, string]>>();
  for (const [method, path] of pairs) {
    const fname = endpointFilename(method, path);
    const arr = inverse.get(fname) ?? [];
    arr.push([method, path]);
    inverse.set(fname, arr);
  }
  const dupes: Array<[string, Array<[string, string]>]> = [];
  for (const [k, sources] of inverse) {
    if (sources.length > 1) dupes.push([k, sources]);
  }
  if (dupes.length === 0) return;
  const lines = ["endpoint filename collisions:"];
  for (const [fname, sources] of dupes) {
    lines.push(`  -> ${fname}`);
    for (const [m, p] of sources) lines.push(`      from ${m.toUpperCase()} ${p}`);
  }
  throw new SlugCollisionError(lines.join("\n"));
}
